Redesign

j-m opened the pull request:
Refactored existing code and then optimised (also looked at the generated js to optimise).
Made constructor args optional.
Added getters.
Added benchmark test.
Removed _ from variable names.
Added semicolons.
Changed tsconfig.
Updated package.json.

j-m committed:
Redesign

Refactored existing code and then optimised (also looked at the generated js to optimise).
Made constructor args optional.
Added getters.
Added benchmark test.
Removed _ from variable names.
Added semicolons.
Changed tsconfig.
Updated package.json.

jmsv commented:
Changing code style and tsconfig was unnecessary - please revert these so meaningful changes are more readable

j-m commented:
Why should `noImplicitAny` be `false`?

j-m committed:
Formatting

Moved one benchmark to separate file and deleted the other.
Removed semicolons.
Formatted.

j-m committed:
fixed the whoops

jmsv commented on index.ts:
```diff
@@ -1,35 +1,65 @@
-  constructor(input: object[], keys: string[]) {
-    this._original = input.map((x) => Object.assign({}, x))
-    this._keys = keys
+  public getKeys(): string[] | undefined {
```
This isn't how getters/setters work - please see https://www.typescriptlang.org/docs/handbook/classes.html

should be something like:
```typescript
get keys(): string[] {
  return this.keys
}
```

j-m replied:
🤯

j-m replied:
Ahhhh now i understand why you need the underscores, silly JS

jmsv replied:
:joy:

j-m replied:
Resolved

jmsv commented on index.ts:
```diff
@@ -1,35 +1,65 @@
+  }
 
-    this._transposed = transpose(this._keys, this._original)
+  public getOriginal(): object[] {
```
should be a getter function - see above comment

j-m replied:
Resolved

jmsv commented on index.ts:
```diff
@@ -1,35 +1,65 @@
 
-  public get(keyName: string, keyValue: string): object[] {
-    return this._transposed[keyName][keyValue]
+  public getTransposed(): object[] {
```
getter

j-m replied:
Resolved

jmsv commented on index.ts:
```diff
@@ -1,35 +1,65 @@
-    }, current || {})
-    return transposed
-  }, {})
+  public push(...data: object[]) {
```
parameter name of `items` was more descriptive imo

j-m replied:
I wanted to standardise the name for it; you used `input` in the constructor.

jmsv replied:
I felt like `input` was more descriptive for `constructor`, and `items` was more descriptive for `push`, but agree it might be a good idea to make them uniform. '`data`' seems generic though

j-m replied:
I've not tested it but my intention is to be able to push objects of different classes, so having a generic name is probably better.

jmsv replied:
but in its current state it only accepts `object[]` - sort of makes more sense to rename when it needs a more accurate name, rather than in anticipation of a feature

j-m replied:
I couldn't think of anything to rename it to where it made sense in both contexts so I left it as `data`.

jmsv commented on index.ts:
```diff
@@ -1,35 +1,65 @@
+    return this.length
+  }
+
+  private getKeysForObject(datum: object): string[] {
```
not sure why we need this

j-m replied:
If options.keys is defined then we want to transpose only those keys, otherwise we'll use the object's.

jmsv replied:
do we reuse this code enough to justify a new function though? 

also, the body of this function does:
```ts
this.keys !== undefined ? this.keys : Object.keys(datum)
```

which is equivalent to:
```ts
this.keys || Object.keys(datum)
```

which is shorter and much more readable

j-m replied:
Man have I missed JS. 
Absolutely changing!

j-m replied:
Resolved

jmsv commented on index.ts:
```diff
@@ -1,35 +1,65 @@
+  }
+
+  private transpose(datum: { [key: string]: any }) {
+    this.getKeysForObject(datum).forEach((key) => {
```
why does the `transpose` use a `forEach` rather that a `reduce` now? I feel like this might have a performance impact

j-m replied:
Read somewhere that reduce, filter, map, and forEach all have roughly the same performance but that the others transform the data, imo using reduce was somewhat misleading.

jmsv replied:
the `reduce` was transforming the array down to a single value: the transformed object - this is exactly the kind of thing reduce is for. would rather this change is reverted

j-m replied:
Right but you had to use two `reduces` to get the same effect as one `forEach`

j-m replied:
Continued in next review

jmsv commented on index.ts:
```diff
@@ -1,35 +1,65 @@
+      if (datum[key] === undefined) {
+        return
+      }
+      if (this.transposed[key] === undefined) {
```
lots of `undefined` checks in this method that weren't needed before - why now?

j-m replied:
I wanted to just do `this.transposed[key][datum[key]] = [datum]` but it throws if `this.transposed[key]` is undefined. 
Same as another comment: If options.keys is defined then we want to transpose only those keys, otherwise we'll use the object's.

jmsv replied:
imo, `=== undefined` is rarely needed, but this relatively short function now uses it 3 times

j-m replied:
Is there a better way to `this.transposed[key][datum[key]].push(datum)` given `this.transposed[key]` or `this.transposed[key][datum[key]]` could be undefined?

j-m replied:
Continued in next review

jmsv commented on package.json:
```diff
@@ -18,13 +18,22 @@
+    "keys",
     "oaty",
     "object",
-    "array",
```
why rm 'array'? 🤔

j-m replied:
I didn't, I alphabetised.

jmsv replied:
oh yeah, y tho

jmsv commented on test/benchmark.js:
```diff
@@ -0,0 +1,57 @@
+var results = []
+
+function time(runs) {
+    var oaty = new OatyArray({data: testArray});
```
no code style enforced for test stuff atm but opting to not use semicolons - think [standardjs](https://github.com/standard)

j-m replied:
I read into ASI and you have to be more careful with JS. I don't think it's worth the risk and that semicolons make it abundantly clear.

jmsv replied:
I've been writing TS/JS without semicolons for a while and haven't had any problems. I'm standing by this 😁

j-m replied:
Okay, I trust you

j-m replied:
Resolved

jmsv commented on index.ts:
```diff
@@ -1,35 +1,65 @@
+  private original: object[] = []
+  private transposed: { [key: string]: any } = {}
+
+  constructor(options?: { data?: object[], keys?: string[] }) {
```
I'm not too keen on `OatyArray` only taking an options object as a parameter. Would prefer the initial array and an `options` object, e.g.

```ts
constructor(input?: array, options?: { keys?: string[] }) {
```

so you can init a new instance as, for example:
```ts
const empty = new OatyArray()
const emptyWithKeys = new OatyArray([], { keys: ['a', 'b'] })
const arrayNoKeys = new OatyArray([...])
```

j-m replied:
As it stands you can omit or redorder the args, like this:
`new OatyArray()`
`new OatyArray({})`
`new OatyArray({ keys: ['a', 'b'] })`
`new OatyArray({ data: [...] })`
`new OatyArray({ data: [...], keys: ['a', 'b'] })`
`new OatyArray({ keys: ['a', 'b'], data: [...] })`

This is more future proof and keeps it customisable for when we add more options.

jmsv replied:
I understand that, but think the `options` parameter should just be for options, rather than options _and_ data.

j-m replied:
Ohhh I'm with you now, agreed.

j-m replied:
Resolved

jmsv commented on index.ts:
```diff
@@ -1,35 +1,65 @@
-    this._transposed = transpose(this._keys, items, this._transposed)
-    return this._original.push(...items)
+  public get(keyName: string, keyValue: string): object[] | undefined {
+    if (this.transposed === undefined
```
loads of undefined checks?

j-m replied:
Same as another comment: If options.keys is defined then we want to transpose only those keys, otherwise we'll use the object's.

But we do need to decide if we want to return something even if it's just an empty array; would make it more consistent and look like it actually tried to search.

jmsv replied:
not sure why `.get` needed the rewrite? imo less readable now

---

> make it look like it actually tried to search

I feel like this is just misleading the consumer of the library - would prefer we returned `undefined`, matching the behaviour of `Array.prototype.find`

j-m replied:
`.get` needs the undefined checks.
Surely `return this._transposed[keyName][keyValue]` would've thrown an NPE if nothing had been pushed or if that key hadn't been transposed?

j-m replied:
`find` also returns a singular object for when there's 1 result rather than an array of 1 object. 
For me at least, consistently returning an array is easier to handle.

j-m replied:
Oh wait I forgot `_this.transposed` is initialised to {} . I'll remove the first undefined check, but I'll have to add one depending on what we return

jmsv replied:
> `find` also returns a singular object for when there's 1 result rather than an array of 1 object.

that's because it searches for the first match, rather than all matches

j-m replied:
oh really? well, that sucks. Soooo, am I returning undefined or [], or is this a new option?

jmsv replied:
return matching object array directly (effectively return `undefined` if there's no match)

jmsv replied:
if we added a function similar to `Array.prototype.filter` in the future, that'd return `[]` if no matches. Our `get` function is more like Array's `find` so returns `undefined`

j-m replied:
Hmmm, would it not make more sense to return `[]` for consistency? Our `get` and `Array.find` are very different, and given we're not calling it `find` I don't think there's any risk of someone expecting `undefined`

jmsv replied:
If I was querying data that doesn't exist, `undefined` is exactly what I'd expect

j-m replied:
Okay, I'll leave it as undefined. I will make it configurable ([], undefined, or throw) in another branch.

j-m replied:
Resolved. Will do as another feature.

jmsv commented on test/benchmark.js:
```diff
@@ -0,0 +1,57 @@
+    return values.reduce((previous, current) => current += previous) / values.length;
+  }
+  
+  time(1)
```
probably better to use a benchmarking/timing library for this and randomised, variable-length datasets

j-m replied:
Yea, just wanted something basic for the mean time.
After this is merged we could create a `profiling` branch and make commit all future integration or memory/time profiling tests against that.

jmsv reviewed (changes requested):
tl;dr:
- use proper get/setters rather than java-esque `getThing()` functions.
- need to sort out better constructor params.
- `transpose`/`get` functions now have loads of `=== undefined` checks and use `forEach` rather than `reduce`, without explanation. if this is for a reason we should discuss it, otherwise revert.

j-m commented on tsconfig.json:
```diff
@@ -5,7 +5,7 @@
     "outDir": "./dist",
     "strict": true,
-    "noImplicitAny": false,
+    "noImplicitAny": true,
```
I feel like having implicit `any`s defeats the point of TS

jmsv replied:
No, just makes types optional. I just made this `false` when I was prototyping but probably a good idea to make it true now

j-m commented on index.ts:
```diff
@@ -1,35 +1,65 @@
-  public push(...items: object[]): number {
-    this._transposed = transpose(this._keys, items, this._transposed)
-    return this._original.push(...items)
+  public get(keyName: string, keyValue: string): object[] | undefined {
```
Also, should we rename this to `find` to be more consistent with other things?

jmsv replied:
Might be misleading - `Array` `find` takes a function parameter, `OatyArray` `get` doesn't

j-m committed:
Code review comments

Readded underscores.
Used `get`s.
Moved `data` out of `options`.
Removed `getKeysForObject`.

Also restructured tests.

j-m commented on index.ts:
```diff
@@ -1,35 +1,60 @@
+  constructor(data?: object[], options?: { keys?: string[] }) {
+    if (data !== undefined) { this.push(...data) }
+    if (options !== undefined) {
+      if (options.keys !== undefined) { this._keys = options.keys }
```
Just realised I can safely remove this undefined check (`options.keys !== undefined`)

jmsv replied:
Go for it :100:

jmsv replied:
just realised, you have an `if` in an `if`, and neither are needed:

```ts
if (options !== undefined) {
  if (options.keys !== undefined) { this._keys = options.keys }
}
```

change this to
```ts
this._keys = options && options.keys || []
```

j-m replied:
Man I love JS.
Will change it to 
`this._keys = options.keys || []`
as we'd be duplicating `options &&` when we add more options

j-m replied:
I'm going to use 
`if (options !== undefined) {
  this._keys = options.keys
}`
Then we won't have to do `options &&` for each new option we have,  and I actually rely on `_keys` being undefined rather than empty for `(this._keys || Object.keys(datum))` in `transpose`.

jmsv replied:
Then stop relying on `_keys` being `undefined` - Was this a problem with the original `transpose` implementation?

j-m replied:
Yes. The new transpose is fundamentally different: I'm adding keys based on the object unless `_keys` is defined, in which case I'm only transposing those.

jmsv replied:
So it sounds like the only difference is how it selects keys?

j-m replied:
Whoops, missed this. Yes, which is why the new version looked had more `if`s than before. 
Outdated.

jmsv commented on index.ts:
```diff
@@ -1,35 +1,60 @@
+  private _transposed: { [key: string]: any } = {}
+
+  constructor(data?: object[], options?: { keys?: string[] }) {
+    if (data !== undefined) { this.push(...data) }
```
these `undefined` checks are still unnecessary and should be rewritten imo, for example, this line can just be:

```ts
this.push(...data || [])
```

j-m replied:
Then I may as well use defaults and do
`constructor(data: object[] = [], options?: { keys?: string[] }) {
    this.push(...data)`

j-m replied:
Imo `if (data !== undefined)` is more readable than having a default

jmsv replied:
I think default params to keep it concise

j-m replied:
Okay

j-m replied:
Outdated. Data is an optional parameter again

jmsv commented on index.ts:
```diff
@@ -1,35 +1,60 @@
   }
 
-  public get(keyName: string, keyValue: string): object[] {
+  public get(keyName: string, keyValue: string): object[] | undefined {
```
this whole function can just be

```ts
return this._transposed[keyName] ? this._transposed[keyName][keyValue] : undefined
```

j-m replied:
Was more in preparation for the future when we'll add different returns (undefined, [], throw) for when an object isn't found.

jmsv replied:
"Premature optimization is the root of all evil" - Donald Knuth

j-m replied:
I agree but it'll be the next thing I do after everything on this review as it's something I need

j-m replied:
Outdated

jmsv commented on index.ts:
```diff
@@ -1,35 +1,60 @@
-    }, current || {})
-    return transposed
-  }, {})
+  private transpose(datum: { [key: string]: any }) {
```
I don't understand the rewrite here and prefer the original; it's more concise and imo more readable. Think `reduce` might also be higher performance than `forEach` used in this rewrite

j-m replied:
reduce is nowhere near as readable; I still have no clue how it work. I don't see how what is essentially is a nested for, could have better performance than just one

j-m replied:
Outdated

jmsv commented on test/test.js:
```diff
@@ -16,44 +16,54 @@ const testArray = [{
-    expect(matches[0]).to.not.be.undefined
-    expect(matches[0].b).to.equal(5)
+  describe('Push', () => {
+    it('push function should return count', () => {
```
usually `describe` and `it` are meant to make sense in terms of grammar - maybe make it something like:

```ts
describe('push function', () => {
  it('should return count', () => {
```

which reads as "push function should return count", for example

j-m replied:
I was going to rename them all when we add more tests to follow "given when then" but I'll do it now

j-m replied:
In fact, I also need to add some more tests so I'll remove the data and use fixtures

j-m replied:
Outdated

j-m commented on index.ts:
```diff
@@ -1,35 +1,60 @@
+  private _original: object[] = []
+  private _transposed: { [key: string]: any } = {}
+
+  constructor(data?: object[], options?: { keys?: string[] }) {
```
I think I may revert back to having `data` inside `options`.
I don't really want to have to specify initialising with an empty array, and I can't do 
`OatyArray({keys: [...]})` because it thinks I'm trying to set data to `{keys:[...]}` which doesn't work

j-m replied:
Reverted

j-m committed:
Code review comments + redesign

Converted tests to typescript and added required dependencies.
Added more npm scripts.
Redesigned tests to use a fixture.
Separated benchmarking from tests.
Added more tests.
Data is optional again.
Tidied .get, ._transpose, and .push.
Added indent enforcement to tslint.

j-m commented on index.ts:
```diff
@@ -1,35 +1,52 @@
+  private _original: object[] = []
+  private _transposed: { [key: string]: any } = {}
+
+  constructor(options?: { data?: object[], keys?: string[] }) {
```
`data` is optional again

jmsv replied:
can you move `data` back outside of `options`?

j-m replied:
Not really, if I wanted to set just the `keys` I would rather call `new OatyArray({keys: ["a"]})` than `new OatyArray([], {keys: ["a"]})`.

jmsv replied:
In most cases an `OatyArray` would be initialised with data - I prefer `new OatyArray(data, {})`

j-m replied:
We should discourage initialising with data, though. 
In my case, I am generating the data. It is much faster to add data when it's created than to add it all at the end. You should only pass in the data if comes from a third party source.

jmsv replied:
> We should discourage initialising with data, though.

Disagree with this - your case isn't everyone's case

j-m replied:
It's not about that. 
`data` cannot be optional if it is outside of `options`.
Passing `null`, `void`, `[]`, `undefined` etc for a mandatory field is bad design; this is the exact purpose of overloads and default/optional arguments.

jmsv replied:
I don't think it's bad design. You initialise an empty array by doing `let arr = []` and I think it makes sense for `OatyArray` to take `[]` as data for empty initialisation.

Regardless, `data` is not an 'option'; the `options` param should be reserved for… options

j-m replied:
I don't agree but I will revert and we can discuss this another time

j-m commented on index.ts:
```diff
@@ -1,35 +1,52 @@
 
   public get(keyName: string, keyValue: string): object[] {
-    return this._transposed[keyName][keyValue]
+    return this._transposed[keyName][keyValue] || []
```
We can make this customisable. For me, it is more useful to return an empty array when nothing is found.

jmsv replied:
alright

j-m commented on index.ts:
```diff
@@ -1,35 +1,52 @@
+    return this._original
   }
 
   public get(keyName: string, keyValue: string): object[] {
```
Given not all data has to have all keys, it may be useful to make `keyValue` optional so that I could get all objects that have that key in O(1), too.

j-m replied:
Given not all data has to have all keys, it may be useful to make keyValue optional so that I could get all objects that have that key in O(1), too.

j-m commented on index.ts:
```diff
@@ -1,35 +1,52 @@
-  public push(...items: object[]): number {
-    this._transposed = transpose(this._keys, items, this._transposed)
-    return this._original.push(...items)
+  public push(data: object[]) {
```
Rest/spread parameters are still new to me; I didn't mean to commit this. 
Are the before and after equivilent?

jmsv replied:
`push`ing without spreading is different, e.g. given:

```ts
const values = [1, 2, 3]
const array = [0]

array.push(values) // [0, [1, 2, 3]]
// or
array.push(...values) // [0, 1, 2, 3]
```

j-m replied:
Ah okay, makes sense. Will revert. Spreading is clearly the more desirable option here

jmsv replied:
yeah `push`ing an array just doesn't make sense

j-m commented on package.json:
```diff
@@ -5,8 +5,10 @@
+    "clean": "rimraf dist",
+    "benchmark": "npm run build && node dist/test/benchmark.js",
+    "build": "npm run clean && tsc",
+    "test": "mocha --require ts-node/register test/**/*.spec.ts --reporter list",
```
Changed the reporter to `list` as `spec` took up too much space imo

jmsv replied:
cool

jmsv commented on dist/test/benchmark.js:
```diff
@@ -0,0 +1,42 @@
+"use strict";
```
why are the tests ending up in the `dist` directory?

j-m replied:
Whoops, didn't commit the .tsconfig `exclude` change

j-m replied:
I had excluded `"**/*.spec.ts"` although this won't exclude `test/benchmark.ts`. I'm not entirely sure where we should put Volumetric & Performance stuff. Thoughts?

jmsv replied:
new directory called `benchmarks` or something?

j-m committed:
Code review comments and several other improvements.

Moved benchmark.
Refactored push & transpose back to how it used to be.
Made an interface for _options.
Moved data outside of _options.
Added config for return types.
Put the initialisers for options and data in the constructor args.
get(key) now returns transposed[key].
Added several tests.

j-m committed:
Alphabetised

jmsv commented on index.ts:
```diff
@@ -1,35 +1,58 @@
-    }, current || {})
-    return transposed
-  }, {})
+  private transpose(data: object[]) {
```
I still prefer the reduce, as before

jmsv reviewed (approved):
looks good :100:

please add an overview of the changes in the readme changlelog under `0.3.0`, and we can merge, bump the version and publish

j-m committed:
Update README.md

j-m committed:
Update README.md
